### PR TITLE
[#13998][Bug][PHP] Move isNullable section to the top of the setter function in templates

### LINKED ISSUE #13998: [BUG][PHP] Runtime exception when property is both nullable and required
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

When a response object contains a property that is:

1. `type: string`
2. `nullable: true`
3. Must conform to some defined `pattern`
4. Is `required` in the object

Then, the generated code for that particular model has the following logic (taken from the generated code in file `lib/Model/Foo.php` of the example below):

```php
     /**
     * Sets bar
     *
     * @param string $bar bar
     *
     * @return self
     */
    public function setBar($bar)
    {

        if ((!preg_match("/^0[xX][a-fA-F0-9]{40}$/", $bar))) {
            throw new \InvalidArgumentException("invalid value for \$bar when calling Foo., must conform to the pattern /^0[xX][a-fA-F0-9]{40}$/.");
        }


        if (is_null($bar)) {
            array_push($this->openAPINullablesSetToNull, 'bar');
        } else {
            $nullablesSetToNull = $this->getOpenAPINullablesSetToNull();
            $index = array_search('bar', $nullablesSetToNull);
            if ($index !== FALSE) {
                unset($nullablesSetToNull[$index]);
                $this->setOpenAPINullablesSetToNull($nullablesSetToNull);
            }
        }

        $this->container['bar'] = $bar;

        return $this;
    }
```
Here, even though `bar` is `nullable`, when `bar == null` there is no check for it before the `isNullable` logic and there is an attempt to `preg_match` the string with the defined pattern, which throws an exception.

##### openapi-generator version

6.2.1

##### OpenAPI declaration file content or url

`phpbug.yaml`
```yaml
openapi: 3.0.2
info:
  title: Bug API
  version: '0.0'
  contact:
    name: Curvegrid
  description: PHP Client Bug API.
servers:
  - url: 'http://{hostname}'
    variables:
      hostname:
        default: localhost:8080
        description: Bug Test Server.
tags:
  - name: test
paths:
  /foo:
    get:
      operationId: foo
      summary: FooBaz
      description: FooBar.
      tags:
        - test
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Foo'
components:
  schemas:
    Foo:
      title: Foo
      type: object
      properties:
        bar:
          type: string
          nullable: true
          pattern: '^0[xX][a-fA-F0-9]{40}$'
      required:
        - bar
```

##### Generation Details

`openapi-generator-cli generate -g php -i phpbug.yaml -o phpclient`

##### Suggest a fix

I noticed looking at https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/php/model_generic.mustache#L388 that the main problem here is that the case where a property can be both `nullable` and `required`, which is completely valid in the specs, isn't handled well in template. I propose that the logic of `isNullable` should be moved to the top of the function, before pattern matching and everything else, _**or**_ the `required` section in lines like https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/php/model_generic.mustache#L435 should be removed so there is an `!is_null` check before trying to match patterns or anything else.

### PR DESCRIPTION
**Pull request by MustansirS**
Resolves #13998 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
@jebentier @dkarlovi @mandrean @jfastnacht @ybelenko @renepardon @wing328

### CODE REVIEW

**Comment by MustansirS** [comment]
@jebentier @dkarlovi @mandrean @jfastnacht @ybelenko @renepardon @wing328 This PR is ready for review

File: modules/openapi-generator/src/main/resources/php/model_generic.mustache [hunk 1]
```diff
@@ -387,10 +387,27 @@ class {{classname}} {{#parentSchema}}extends {{{parent}}}{{/parentSchema}}{{^par
...
+        {{^isNullable}}
+        if (is_null(${{name}})) {
+            throw new \InvalidArgumentException('non-nullable {{name}} cannot be null');
+        }
+        {{/isNullable}}
         {{#isEnum}}
         $allowedValues = $this->{{getter}}AllowableValues();
         {{^isContainer}}
-        if ({{^required}}!is_null(${{name}}) && {{/required}}!in_array(${{{name}}}, $allowedValues, true)) {
+        if ({{#isNullable}}!is_null(${{name}}) && {{/isNullable}}!in_array(${{{name}}}, $allowedValues, true)) {
```
[the remark this answers is not in the document]
Comment (MustansirS) [reply modules/openapi-generator/src/main/resources/php/model_generic.mustache:410]: I noticed that if it is an optional field, a `null` value will never be passed to the `setter` fucntion

[the remark this answers is not in the document]
Comment (MustansirS) [reply modules/openapi-generator/src/main/resources/php/model_generic.mustache:410]: @wing328

[the remark this answers is not in the document]
Comment (wing328) [reply modules/openapi-generator/src/main/resources/php/model_generic.mustache:410]: thanks for the explanation.

**Comment by wing328** [comment]
tests passed via https://app.travis-ci.com/github/OpenAPITools/openapi-generator/builds/257997016

**Comment by wing328** [comment]
thanks for the fix. lgtm but i've not tested it locally

let's give it a try to see if anyone has further feedback on this fix.

**Comment by MustansirS** [comment]
Thank you!

**Pull request merged**